sale_quotation_number/models/sale_order.py

```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+# © 2010-2012 Andy Lu <andy.lu@elico-corp.com> (Elico Corp)
+# © 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+# © 2017 valentin vinagre  <valentin.vinagre@qubiq.es> (QubiQ)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from odoo import models, api, fields
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    keep_name_so = fields.Boolean(
+        string="Use Same Enumeration",
+        help="If this is unchecked, quotations use a different sequence from "
+             "sale orders", default=True)
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    keep_name_so = fields.Boolean(
+        related='company_id.keep_name_so',
+        readonly=False
+    )
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.multi
+    def copy(self, default=None):
+        self.ensure_one()
+        if default is None:
+            default = {}
+        default['name'] = '/'
+        if self.origin and self.origin != '':
+            default['origin'] = self.origin + ', ' + self.name
+        else:
+            default['origin'] = self.name
+        return super(SaleOrder, self).copy(default)
+
+    @api.model
+    def create(self, vals):
+        company = self.env['res.company']._company_default_get('sale.order')
+        if not company.keep_name_so:
+            vals['name'] = self.env['ir.sequence'].next_by_code(
+                'sale.quotation') or '/'
+        return super(SaleOrder, self).create(vals)
+
+    @api.multi
+    def action_confirm(self):
+        if super(SaleOrder, self).action_confirm():
```

Comment:
@manuelregidor 

There is a small problem with this process.

If you confirm the order, will put the number on the procurement, and will be the reference on Manufacturing Orders.

If you then change the number after it is confirmed, it will not match the procurement references.

We found you really need to allocate the number and put it on the order, and then call action confirm, so that the downstream references match the new number, not the quote number.

Reply:
I have tried to sort out that issue. You can check it out in my last commit. 

However, there are still a few problems. Everything works fine as long as just one SQ is confirmed. If multiple SQs get confirmed altogether, the value of number_next_actual (ir.sequence model), which I use, does not get updated until all the SQs have been confirmed.

Maybe you can check it out and find out a way of solving the problem. Otherwise, I will just return to the way the module works in previous versions.

Reply:
@manuelregidor Seems very convoluted - confirm always returns true, unless it raises an error, in which case it should roll back.


`
    @api.multi
    def action_confirm(self):
        for order in self:
            if order.state in ('draft', 'sent') and not order.company_id.company.keep_name_so:
                if order.origin and order.origin != '':
                    quo = order.origin + ', ' + order.name
                else:
                    quo = order.name
                order.write({
                    'origin': quo,
                    'name': self.env['ir.sequence'].next_by_code(
                        'sale.order')
                })
        return super().action_confirm()
`

Should be sufficient - if any order can't confirm it will roll back?

Reply:
If an order can't confirm, the sequence will go forward anyway and there will be gaps.

Reply:
@manuelregidor 

No, if an order can't confirm, it raises an error.

The code in sale.py returns True every time, unless it raises an error on the way.

I am happy to be shown wrong, but can you point out the code in Odoo that returns anything other than True (or the result of self.write, from action_done, which is always True, too).

Reply:
@richard-willdooit 
Yes, it always returns True inless it raises an error. That's why, when an order can't confirm because an error is raised in action_confirm (from sale_order parent), next_by_code has already been done, so the code has moved to the next number. Next time an order is succesfully confirmed, there will be a gap in the sequence.

Reply:
@manuelregidor No, when an error is raised, the transaction rolls back, and the increment did not happen.  The sql issues a rollback.

Reply:
@manuelcalerosolis The increment of the number is inside the same sql transaction

Reply:
@richard-willdooit Yes, you're right. I've tested it with "real" errors (not forcing exceptions as I was doing before) and it works fine. Thank you.